### ISSUE TITLE
Add RPCs for vessel MoI and torque

### ISSUE THREAD

**Issue by djungelorm** [krpc/krpc#212]
- [ ] Vessel.MomentOfInertia - moment of inertia of the vessel
- [ ] Vessel.Torque - total torque. Sum of RCS, ReactionWheel and CurrentGimballedEngine torques
- [ ] Vessel.RCSTorque - torque from RCS
- [ ] Vessel.ReactionWheelTorque - torque from active reaction wheels
- [ ] Vessel.GimballedEngineTorque - torque from currently active engines, at the current throttle setting

**Comment by lamont-granquist** [comment]
I started playing with this a bit and I think MOI is the biggest thing to export.

It looks like KSPs MOI is pretty horribly buggy and may not include the mass of the resources in a part.

Here's is what MechJeb does and its reasonably understandable once I start dusting off some old braincells from physics class:

https://github.com/MuMech/MechJeb2/blob/ad7371b08ce62edc3877b965ca5b2054115aff29/MechJeb2/VesselState.cs#L914-L965

I started down this road by thinking about adding Part.InertiaTensor:

```
diff --git a/service/SpaceCenter/src/Services/Parts/Part.cs b/service/SpaceCenter/src/Services/Parts/Part.cs
index e9c273a..915d916 100644
--- a/service/SpaceCenter/src/Services/Parts/Part.cs
+++ b/service/SpaceCenter/src/Services/Parts/Part.cs
@@ -159,6 +159,14 @@ namespace KRPC.SpaceCenter.Services.Parts
         }

         /// <summary>
+        /// The inertia tensor of the part.
+        /// </summary>
+        [KRPCProperty]
+        public Tuple3 InertiaTensor {
+            get { return InternalPart.rigidbody ? InternalPart.rigidbody.inertiaTensor : new Tuple3 (0, 0, 0); }
+        }
+
+        /// <summary>
         /// The impact tolerance of the part, in meters per second.
         /// </summary>
         [KRPCProperty]
```

I got that far, and then tried to compile krpc on a Mac and wound up landing in a field of poorly shaven yaks armed only with a dull razor blade...

I'll need to setup an Ubuntu virt for development work...

**Comment by djungelorm** [comment]
Ah yes, the bazel build scripts won't play well on Mac. I've added an archive to the release page containing the generated files and libraries that you need to build the DLLs using the KRPC.sln solution file. This should let you play around with KRPC.SpaceCenter.dll without needing to set up a linux build env.

https://github.com/krpc/krpc/releases/download/v0.2.2/krpc-genfiles-0.2.2.zip

Just extract that archive over a clone of the repo. As you're adding RPCs, these won't show up in the C#/C++/Java clients (you'd need to regenerate some of the code provided in that zip) but they will be visible to the python and lua clients (they dynamically query the server to get the available RPCs).

**Comment by lamont-granquist** [comment]
Yeah, I hit a problem where the version of bazel on homebrew was too old, upgraded it from the homepage, then had to hack up a KSP_Data symlink, then Macs don't have an '-r' option for 'ln' so I installed coreutils and changed all the 'ln' calls to 'gln' then something else blew up on me, and i decided that was far enough...

I should really be able to get an Ubuntu virt going fairly quickly, and automate most of the setup but nothing seems to be working for me today...

**Comment by djungelorm** [comment]
The build scripts are a bit of a nightmare - I'd recommend not looking in the tools/build folder... But Bazel is far better than the custom Makefiles I used to use!

The [travis config](https://github.com/krpc/krpc/blob/master/.travis.yml) is a good place to look if you need hints on how to get the build working.

**Comment by lamont-granquist** [comment]
Are you using Ubuntu 14.04 or 15.10?  I got it building on 14.04 and got 4 tests passing, but got some that have some odd looking errors and makes me think mono on 14.04 might be out of date or something...

**Comment by djungelorm** [comment]
I'm on 15.10. Travis CI is running 14.04 and that build things OK.

On both systems, I think you need to install the latest version of mono: http://www.mono-project.com/docs/getting-started/install/linux/#debian-ubuntu-and-derivatives

Also the version of luarocks in the 14.04 apt repositories is too old, so I had to install it manually on Travis: https://github.com/krpc/krpc/blob/master/tools/travis-ci/install-luarocks.sh

**Comment by djungelorm** [comment]
I've been having a look at MechJeb's VesselState code for computing torques, and have identified a bunch of RPCs that if added would allow torque calculations to be done in the client - e.g. getting the thrust vectors of engines, RCS thrusters etc. I've added them to the list in the first post.

**Comment by djungelorm** [comment]
I've coded up the interface for the new RPCs, but most of them just throw NotImplementedException at the moment. If you want to have a go at implementing any of them feel free! Just post on here so we don't duplicate our efforts. I'm currently working on implementing the thrust direction reference frame.

**Comment by djungelorm** [comment]
The more I think about it, I don't think we need Engine.InitialThrustDirection (thrust direction ignoring gimballing). And if you really need it you could calculate it from Engine.ThrustDirection and Engine.GimbalRotation

**Comment by lamont-granquist** [comment]
* still working on build box, haven't had a chance to update mono or luarocks yet, that's most likely the problem..

**Comment by lamont-granquist** [comment]
Okay, so the MOI inside of kerbal is terribad.

It just sums over all the parts and takes its `drymass * radius^2` from the CoM of the vehicle as a pointmass.

That means a single part has a moment of inertia of `[0,0,0]`

A vehicle that is just a stack of two Jumbo64s has a moment of inertia of `[112.5, 0, 112.5]`.   There are at least three problems with this:

- the MOI around the y-axis is zero
- the MOI around the x-axis and z-axis is too low due to the pointmass approximation (around 112.5 vs 162.5)
- the MOI around the x-axis and z-axis are waaaaay too low due to using the dry mass (and really should be around 1462.5 taking into account the fuel)

So `[ 1462.5, 225, 1462.5 ]` is closer to what it should be for a stack of two jumbo tanks (in t/m^2 so *1000f for kg/m^2)

**Comment by djungelorm** [comment]
Is that Vessel.findLocalMoI that you were testing?

**Comment by djungelorm** [comment]
I've an implementation for MoI, calculated using Vessel.findWorldCenterOfMass and the rigidbody objects for each part.

btw, I'm pushing changes related to this issue to the branch feature/autopilot-rpcs

**Comment by lamont-granquist** [comment]
Yeah, I've been building off your code.  Your MOI code looks correct for what it does, but its still a pointmass approximation.  However, the numbers out of it also look a bit off, and I'm suspecting the vectors aren't in the right basis.  My jumbo-64 stack gives me '[326.50346068853514, 1012.5020078035905, 686.0043089140931]' which just isn't symmetric around the y-axis for one thing.

**Comment by djungelorm** [comment]
OK, my assumption that part.rb.position is in world space is probably wrong in that case.

Must admit my physics is a little lacking - but I would have though KSPs physics simulation is using a pointmass model? Or does it do something more fancy?

**Comment by lamont-granquist** [comment]
Yeah, I think it is in worldspace, but you need it in the vessel's coordinate system.  We want to get the pitch, roll, yaw components of the Inertia tensor relative to the vessel.  I did this and it worked:

```cs
var position = InternalVessel.GetTransform().InverseTransformDirection(part.rb.position - CoM);
```

With that I get `[1012.5054717165618, 4.089142977350968e-08, 1012.5054717518974]` which looks correct for the pointmass approximation.

**Comment by lamont-granquist** [comment]
^ that transform is just copypasta out of what MJ does

**Comment by djungelorm** [comment]
Ah yes of course, makes sense.

Looking at https://github.com/MuMech/MechJeb2/blob/master/MechJeb2/VesselState.cs#L917
there is mention of part.rb.worldCenterOfMass. I wonder how this differs to part.rb.position...

**Comment by lamont-granquist** [comment]
So KSPs physics simulation actually uses the inertia tensors of all the objects.  It has to, because if you build a rocket out of a single stack of parts with no wings or anything then you'd have a moment of inertia that looked like `[ X, 0, Z ]` which would mean you'd be 'massless' around the y-axis so any tiny torque applied to roll the rocket would spin it up with a massive angular velocity.

What we've got now is probably good for PID controllers for launching rockets since 1012 is close to 1462 and what you care about there is taking a stack of parts and gradually pitching it over into a gravity turn...

**Comment by lamont-granquist** [comment]
And yeah, this bit is basically what you did tonight:  https://github.com/MuMech/MechJeb2/blob/master/MechJeb2/VesselState.cs#L947-L959

Although MJ calculates the full inertia tensor and then selects only the diagonal elements.  Although it looks like it might use the off-diagonal elements to compute the angular velocity:  

https://github.com/MuMech/MechJeb2/blob/82b9a3ab4d98f8248d6bf3d5827e8436fd2414b7/MechJeb2/VesselState.cs#L963

The bit you are missing is this part:

https://github.com/MuMech/MechJeb2/blob/82b9a3ab4d98f8248d6bf3d5827e8436fd2414b7/MechJeb2/VesselState.cs#L933-L945

**Comment by lamont-granquist** [comment]
(also if you want kg/m^2 need to multiply by 1000, you're getting t/m^2 right now)

**Comment by djungelorm** [comment]
Ah yes makes sense. Should have just copied MechJeb in the first place! We should also add a Part.InertialTensor RPC so people can access them directly.

And yes I like to stick to SI units to keep the API nice and clean :)

**Comment by djungelorm** [comment]
And probably also worth adding Vessel.InterialTensor so people can get hold of the full thing

**Comment by lamont-granquist** [comment]
It'd be kinda nice to know what MJ is really doing, and actually grok the code.  So far I like your MOI calculation better than what MJ is doing since there's fewer calculations and you throw less away...

And yeah, Part.InertiaTensor, Part.inertiaTensorRotation, Part.CenterOfMass, and possibly Part.CenterOfMassPosition.

I wrote a bit of code that at least compiles for the latter two that I was gonna play a bit with:

```cs
        /// <summary>
        /// The center of mass of the part in the given reference frame.
        /// </summary>
        /// <param name="referenceFrame"></param>
        [KRPCMethod]
        public Tuple3 CenterOfMass (ReferenceFrame referenceFrame)
        {
            return referenceFrame.PositionFromWorldSpace (InternalPart.rigidbody.worldCenterOfMass).ToTuple ();
        }

        /// <summary>
        /// The position of the part.
        /// Measured from the center of mass of the vessel to the center of mass of the part.
        /// </summary>
        /// <param name="referenceFrame"></param>
        [KRPCMethod]
        public Tuple3 CenterOfMassPosition (ReferenceFrame referenceFrame)
        {
            return referenceFrame.DirectionFromWorldSpace (InternalPart.rigidbody.worldCenterOfMass - InternalPart.vessel.findWorldCenterOfMass ()).ToTuple ();
        }
```

**Comment by lamont-granquist** [comment]
anyway, i gotta crash tonight, i'll check in tomorrow and see how far you got...

**Comment by djungelorm** [comment]
By "copy" I meant read, understand and reimplement. Will add those RPCs to the ever growing list...

I need to crash too. Night. And thanks for the help :)

**Comment by lamont-granquist** [comment]
I've almost managed to understand what MJ is doing with the inertia tensor now.

The rigidbody.inertiaTensor is actually the diagonalized inertiaTensor along the principle axis of rotation of the part.  The rigidbody.inertiaTensorRotation is the rotation that would be applied to get the inertia tensor in the reference frame of the part.  Probably for most parts the inertiaTensorRotation is going to be the identity rotation since the parts will be symmetric.  At any rate those values don't change during the game and are describing the geometry of the part.   Then Mechjeb is applying the part rotation, then it is applying the transform from world coordinates to vessel coordinates (the inverse vessel transform).

So that explains this line:

```cs
Quaternion princAxesRot = inverseVesselRotation * p.transform.rotation * rigidbody.inertiaTensorRotation;
```

Then while you apply a transformation to a vector like `v' = T v' you apply a transformation to a tensor like `I' = T^-1 I T` so you need the inverse of that transform.

```cs
Quaternion invPrincAxesRot = Quaternion.Inverse(princAxesRot);
```

And then Unity doesn't seem to have a 3x3 matrix, and you can only apply quaternions to vectors ('T v') there's no operator to apply a vector to a quaternion ('v T'), adding that on top of the fact that the inertia tensor we're manipulating is just a vector of the diagonal elements, and I think you recover all the complexity of:

```cs
                for (int j = 0; j < 3; j++)
                {
                    Vector3 partInertiaTensorTimesjHat = princAxesRot * Vector3.Scale(principalMoments, invPrincAxesRot * unitVectors[j]);
                    for (int i = 0; i < 3; i++)
                    {
                        inertiaTensor[i, j] += Vector3.Dot(unitVectors[i], partInertiaTensorTimesjHat);
                    }
                }
```

Which I think is just `I' = T^-1 I T` but I'm still digesting that one a bit.

**Comment by djungelorm** [comment]
I must admit the math/physics is a bit beyond me... I'm going to look into implementing some of engine gimbal/rcs thrust vectors stuff while you get your head around the math.

**Comment by lamont-granquist** [comment]
Did some work this morning, gotta run now.  I think this is actual working InertiaTensor code:

https://github.com/lamont-granquist/krpc/commit/03ad721abb00da6a30bf367fcf2affbc30af1825

The thing I like about this is that its real obvious that I'm applying the parallel axis theorem and the math is clear.  On the CPU its probably not very well optimized, but I think by reutilizing the 4x4 matrix class from Unity that when KSP 1.1 + PhysX 3.3 drops that the code will start executing on the GPU.

Its a little bit stupid because all Unity has is a 4x4 matrix and it doesn't have basic operators like '+' defined on it.

Still lots of debugging in the code.  I might try creating a utility matrix class that wraps and extends the Matrix4x4 class--those utility methods in the vessel class are not in the right place.  I validated that with a single Jumbo-64 that I recovered the inertiaTensor of the part itself.  With two of them it looks right (in particular rotation around the y-axis is doubled from the single part, the x and z axes are much higher and close to what I computed by hand -- I think the inertia tensor of the jumbo-64 part itself might be a bit off).

**Comment by djungelorm** [comment]
Cool. Thoughts:
 * Do we really need Part.CenterOfMassPosition? Can just be computed from Vessel.CenterOfMass and Part.CenterOfMass.
 * I like how you've emulated 3x3 matrices using 4x4 matrices!
 * There is a GeometryExtensions class that these tensor/matrix routines could be added to.

**Comment by djungelorm** [comment]
Other thoughts I've had:

 * Currently there is Vessel.Position (CoM of the vessel) and Part.Position (arbitrary position, defined by the part model). Maybe we should add:
  * Vessel.CenterOfMass as an alias of Vessel.Position
  * Add Part.CenterOfMass and keep Part.Position as it is. I think it's useful to having a position that isn't affected by fuel drain etc.
  * Have two part reference frames: one with its origin at the CoM of the part, and one at its Part.Position
 
I've also committed some changes for engine/rcs thrust transforms. Engine and RCS objects now have a 'Thrusters' property that returns a list of Thruster objects. These contain the thrust direction and position. These thruster lists are needed because Engines/RCS can have multiple nozzels (e.g. the RAPIER engine with its four nozzels, or the 4-way RCS block).

**Comment by lamont-granquist** [comment]
"Do we really need Part.CenterOfMassPosition?"  -- probably not, that was back when I was thinking about people constructing the inertiaTensor or MoI in their own code.

"There is a GeometryExtensions class that these tensor/matrix routines could be added to." -- I'll check it out

Not sure about having two different reference frames of the part.  I think the CoM frame would be pretty lightly used outside of the inertiaTensor computation for the whole vessel..

**Comment by mthiffau** [comment]
Hi Lamont, looks like we've been working in parallel for the past couple days. I've also come to the conclusion that findLocalMOI seems to be omitting the per-part inertia tensors (and possibly also resource values). I actually started off looking at the MoI calculations being done by FAR in the stability derivative panels in the VAB/SPH. 

I am in a position of understanding the point mass approximation code rather well, but I don't quite grok what is being done to deal with the per-part inertia tensor contributions. I've also implemented a new MoI function by adapting the FAR code to run in a partmodule (using some code picked from the MJ implementation to deal with reference frames). 

The code Ferram wrote looks pretty straight forward (despite not understanding what it does), and I generally trust his competance. The code he wrote is here: https://github.com/ferram4/Ferram-Aerospace-Research/blob/master/FerramAerospaceResearch/FARGUI/FAREditorGUI/Simulation/StabilityDerivCalculator.cs

I also have a copy of the MJ implementation copied verbatim. However the MJ code and the adapted FAR code disagree in flight. I'm not sure if the FAR code is wrong because of some assumption made about coordinate systems in the editor (which it was made for) that I haven't found and fixed yet, but when I checked the part positions and rotations it was generating they were correct. 

I'm interested to see how the adapted FAR code output compares to your new code's results, if your code disagrees with the original MJ code also. 

Just a note: FAR's calculated MoI values in the VAB are *WRONG*. Though it may not be due to the math FAR is using. The point-mass approximation part of his code agrees with the MJ code, it's the inertia tensor part that's different. However I noticed FAR's inertia tensor calculation disagreed with itself between in-game and in the editor. Turns out part.rb.inertiaTensor returns (1,1,1) for a bunch (but not all) of the part inertia tensors in the editor. I'm not sure if that's a unity bug, a KSP bug, or what, but the FAR algorithm can't generate correct MoI's with garbage data whatever the case.

**Comment by mthiffau** [comment]
https://github.com/mthiffau/FARKOSData/blob/stock_far/FARKOSData/FARKOSData.cs

That has the implementations I've been working with so far. The mod is just a little one-part mod I created to have a custom part module to give me data I wanted for use with kOS/kRPC. for things there aren't RPC's/kOS functions for yet. 

the functions are called calculateMOI and calculateMOI2. The calculateMOI3 is just using the findLocalMOI for comparison.

**Comment by lamont-granquist** [comment]
So mine also works:  https://github.com/lamont-granquist/krpc/commit/03ad721abb00da6a30bf367fcf2affbc30af1825

I /think/ its a better implementation than the others floating around since by virtue of reusing the Matrix4x4 it'll wind up running on the GPU when KSP 1.1.0 slurps in Unity 5 and PhysX 3.3

That makes sense in my head anyway (I can't imagine the Matrix4x4 class in Unity 5 wouldn't be built on the GPU-accelerated PxMat44 class).   I asked a question on the Unity forum over here about it:

http://forum.unity3d.com/threads/unity-3x3-matrix-class-with-gpu-accelleration.390171/

**Comment by lamont-granquist** [comment]
The parallel axis theorem is just this:

![screen shot 2016-03-07 at 9 19 33 am](https://cloud.githubusercontent.com/assets/454857/13577130/cf002d4c-e445-11e5-8fb2-2cf4c1517a88.png)

The added complexity comes because the basis of the rigidbody.inertiaTensor isn't in the basis of the vessel, so you have to transform the tensor `I` there as `O-1 I O` where O is the combined operation of transforming from the inertiaTensor's frame to the parts frame to the world frame then to the vessel frame (so you have 3 transforms that you multiply then you take the inverse).

**Comment by lamont-granquist** [comment]
I'm not certain what ferram is doing or how he's implemented it.  I'm certain that its terribly efficient running on the CPU but its difficult to peel away the optimization he's done.  That's kind of why I like the approach of re-utilizing the Matrix4x4 class.  The math remains really clear, and the optimization should be done (if i'm right) by feeding it to the GPU which will just chew up each vector operation in one gulp...  Wish we had KSP 1.1.0 already so I could benchmark...  I'm certain the matrix class is going to be slower on 1.0.x

**Comment by mthiffau** [comment]
As we speak I'm firing up KSP to run your implementation and the one I adapted from his side by side. I really hope they agree and we can put this to bed :P

**Comment by lamont-granquist** [comment]
So, I just realized a much faster implementation would be to check and see if mechjeb or FAR or KoS or any of the other mods out there that do this computation on every tick are installed and just get and return the value from them...

Really this is screaming out for a common library...

Also when I get a chance later today I'll write some tests to make checking algorithms a bit easier...

**Comment by mthiffau** [comment]
So I have some bad news unfortunately. Like I said I pulled the implementation you linked to and put it into my custom PartModule code so I could compare it to the one I adapted from FAR. For a relatively normal looking airplane, the FAR implementation is giving me (1, 32, 39) t/m^2 where yours is giving (0, 6, 48) t/m^2. I wouldn't say for sure that yours was wrong, except that I'm certain the MoI is more than zero for the craft roll axis.

**Comment by lamont-granquist** [comment]
Heh, cool.  Well for something I whipped up yesterday I wouldn't be too surprised if it had a bug still lurking in it.  The only thing I really checked analytically is that for a stack of two jumbo-64s that it got the y-axis right since that one has no displacement from the parts y-axis and the mass (and therefore the inertia) is just doubled.  That doesn't check the rotation from the part frame to the vehicle frame, or either of the computations with an offset axis...  We need some decent tests.

**Comment by mthiffau** [comment]
Indeed. On the bright side, looking through your code and the stuff that isn't implemented yet, some of the other hard stuff is already done by FAR (ie control surface torque) and can likely be adapted to run in real time. 

If you want to go over a test plan later and figure out what we need to do to actually figure out whats working, I'm happy to help. I teach at a program for kids that's kind of like scouts/guides, so I'll be gone from 5:30 to 10:00 PM EST this evening but I'll be around after. Is there a KRPC IRC or slack channel we can chat on?

**Comment by lamont-granquist** [comment]
Yeah, this whole issue kinda reads like a chat log, and I've been wondering about that.  I'm already on 4 slack orgs, so I'd tend to prefer that....

**Comment by mthiffau** [comment]
Sure, I'm on the kOS slack channel. Is there one for kRPC?

**Comment by lamont-granquist** [comment]
is there a kerbal slack org with a #kos channel?  or is it a kos-specific slack org?  either way you can send me an invite to lamont@scriptkiddie.org

**Comment by lamont-granquist** [comment]
I got the inertia tensor computation working last night with @mthiffau 
I rewrote it today and moved a lot of the code to GeometryExtensions and removed the debugging (have not tested the refactoring)
I also added reaction wheel torque (have not tested)

**Comment by djungelorm** [comment]
Awesome job :) Do you want me to merge the commits into the feature/autopilot-rpcs branch?

In other news, I've added a bunch of stuff to do with engine gimbals - so it should now be possible to compute the available engine torque from an external script.

**Comment by djungelorm** [comment]
Also, I've not thoroughly tested my changes so far. Once the new API has settled down a bit I'll write some more exhaustive tests. (FYI I have a bunch of python scripts to automate testing [here](https://github.com/krpc/krpc/tree/feature/autopilot-rpcs/service/SpaceCenter/test))

**Comment by lamont-granquist** [comment]
Yeah I looked at your python tests and couldn't quite figure out how to run them.  You had a hardcoded 10.x.x.x address that I changed to 127.0.0.1 which got it connecting, but it was blowing up somewhere else, I was meaning to ping you about that...  

Do either of you know how to convert the angular velocity into pitch/yaw/roll acceleration?  I was playing with writing a double-PID controller for pitch yesterday and the last thing I need is the angular pitch velocity...

And yeah, you can pull my branch into yours now, the MOI and the RW torque are working.

Oh, MOI needs to be multiplied by 1000f still though...   The RW code re-uses yours so its already got that factor in it...

**Comment by djungelorm** [comment]
You also need TestingTools.dll in your GameData directory. Build it using ```bazel build //tools/TestingTools``` and copy ```bazel-bin/tools/TestingTools/TestingTools.dll``` to GameData. (I use the script ```tools/install.sh``` to build everything and copy it to GameData so I don't have to do this manually)

**Comment by lamont-granquist** [comment]
heh, watching the tests run right now, that's pretty cool...

**Comment by djungelorm** [comment]
Yeah it's pretty nifty :) I've merged your changes - thanks again.

**Comment by djungelorm** [comment]
I've been learning control theory and have implemented a prototype roll controller that auto-tunes tuning using the new reaction wheel torque and MoI functionality. More info here: http://forum.kerbalspaceprogram.com/index.php?/topic/62902-105-krpc-remote-procedure-call-server-v022-28th-feb-2016/&do=findComment&comment=2460919

**Comment by djungelorm** [comment]
I have merged the changes so far into the master branch 9fd7ec91 and they will be included in version 0.3.0.

**Comment by lamont-granquist** [comment]
@djungelorm you should test the 1.1 prerelease's vessel.MOI it may be fixed now -- the code we built here got included into 1.1

**Comment by djungelorm** [comment]
Ah cool, although I literally just pushed the tags for v0.3.0!

Will have to wait till 0.3.1

**Comment by djungelorm** [comment]
Looks like it checks out. I have an example vessel and the values it returns for that match our implementation.

**Comment by lamont-granquist** [comment]
slickness.  i kinda want to get some time to work on this again and write some tests to exercise some edge conditions for it...

also there's a new ITorqueProvider API (attached to every part?  every relevant part?) that gives possible torque from reaction wheels, engines, control surfaces, etc as a (pitch, roll, yaw) vector.

**Comment by djungelorm** [comment]
Ooo that's very interesting, I was slightly dreading having implement gimbal/rcs/aero torques manually.

**Comment by lamont-granquist** [comment]
Yeah I think @nathankell fixed that for you...

**Comment by djungelorm** [comment]
I have been doing some testing, and the `GetPotentialTorque` methods for reaction wheels, RCS, engines and control surfaces give the value we want :)

Only thing I haven't figured out is how to get the vessels full moment of inertia tensor from the vessel. Vessel.MoI gives the diagonal elements. This isn't a major issue though as the custom code you wrote computes it correctly.

**Comment by lamont-granquist** [comment]
yeah the full inertia tensor isn't something that people should need very much, if ever -- the moment of inertia is just the diagonal elements which give you pitch, roll, yaw which is sufficient for most purposes in writing pid controllers (the torque for RCS/etc will be in the same frame as well).  for MOI you should probably return Vessel.MOI.  you could keep the inertiaTensor method around if anyone wants it, but i kind doubt they'll ever need it...

**Comment by djungelorm** [comment]
I've been doing some testing of KSPs Vessel.MoI and it looks like it's still a bit broken. (Not sure how I ever got the output to agree with the custom code a few comments back - maybe I was low on sleep and messed up...)

With a single FLT-400 fuel tank, KSP computes the MoI as ```879.92, 879.92, 441.48```. Looks like the roll and yaw values are swapped? Compare this to the custom code which returns ```879.92, 441.48, 879.92```.

It gets worse... with two FLT-400 fuel tanks stacked on top of each other, KSPs MoI is ```5728.39, 5728.39, 4851.51```. The roll component is now 4850 (assuming we need to swap roll and yaw) but this looks wayyyy too high. The custom code returns ```5728.39, 882.96, 5728.39``` (with 882 for the roll value which is double the value for a single tank - makes sense).

Maybe the MoI is in some other reference frame? Not sure how you'd be able to convert it though as it's not really a vector quantity?

I also checked the latest MechJeb code - they still use custom code. And kOS uses ```Vessel.findLocalMoI(Vessel.findWorldCenterOfMass)```. I tried that but it always returns ```0,0,0``` Do you have any ideas? Maybe we should just use the custom MoI code?

**Comment by lamont-granquist** [comment]
You should probably ship with the custom code for now if it works better.

There was some argument over that, the kOS author thought it needed the additional rotation applied to it, and that seems to work for kOS.  I didn't have time to test any of it before 1.1.0 shipped, so it went out the door with the rotation applied.  I was a little worried it smelled wrong, but so far nobody complained yet.

I'll take a look tomorrow and see what I can figure out.

**Comment by djungelorm** [comment]
OK thanks! The math for MoI is a bit beyond me :P Also just noticed that there is a parameterless version of ```Vessel.findLocalMoI``` but it returns garbage. ```205.71, 0.10, 205.82``` for a single FL-T400.

**Comment by lamont-granquist** [comment]
ah @nathankell says its fixed in git already so 1.1.3 should be better

**Comment by djungelorm** [comment]
OK, will keep my eye out when 1.1.3 drops

**Comment by v0lat1le** [comment]
Would it be possible to add ability to get torques imposed by aero forces (possibly excluding control surfaces)? Should make more precise autopilots  easier.

**Comment by djungelorm** [comment]
Closing as completed. For aero forces, see #914
